### ISSUE TITLE
SAM2DynamicInteractivePredictor fails when max_obj_num=1 due to obj_id_to_idx mapping

### ISSUE THREAD

**Issue by pittforce** [ultralytics/ultralytics#22528]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

When using SAM2DynamicInteractivePredictor with max_obj_num=1, the call to predictor() fails because the internal _obj_id_to_idx() function does not handle the single-object case properly.

The issue occurs when obj_ids=[0] or obj_ids=[1] is passed to the predictor — the lookup fails, leading to an exception since obj_id_to_idx mapping returns None.

This seems unintended, as single-object segmentation is a valid use case for SAM2.

### Suggested Fix

In models/sam/predict.py (around line 1982), update _obj_id_to_idx() to handle the single-object case:

```
def _obj_id_to_idx(self, obj_id: int) -> int | None:
    """Map client-side object id to model-side object index."""
    if self._max_obj_num == 1:  # Handle single-object scenario
        return 0
    return self.obj_id_to_idx.get(obj_id, None)
```

This ensures consistent mapping and allows the predictor to work properly when only one object is tracked.

### Environment

Ultralytics 8.3.222  Python-3.11.13 torch-2.9.0+cu128 CUDA:0 (NVIDIA GeForce GTX 1650 Ti with Max-Q Design, 4096MiB)   
Setup complete  (8 CPUs, 15.8 GB RAM, 463.0/475.7 GB disk)

OS                     Windows-10-10.0.26100-SP0
Environment            Windows
Python                 3.11.13
Install                pip
Path                   C:\Users\janpe\miniconda3\envs\dlc\Lib\site-packages\ultralytics
RAM                    15.76 GB
Disk                   463.0/475.7 GB
CPU                    11th Gen Intel Core(TM) i5-1135G7 2.40GHz
CPU count              8
GPU                    NVIDIA GeForce GTX 1650 Ti with Max-Q Design, 4096MiB
GPU count              1
CUDA                   12.8

numpy                   1.26.4>=1.23.0
matplotlib              3.8.4>=3.3.0
opencv-python           4.11.0.86>=4.6.0
pillow                  11.3.0>=7.1.2
pyyaml                  6.0.3>=5.3.1
requests                2.32.5>=2.23.0
scipy                   1.16.2>=1.4.1
torch                   2.9.0+cu128>=1.8.0
torch                   2.9.0+cu128!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision             0.24.0+cu128>=0.9.0
psutil                  7.1.1
polars                  1.34.0
ultralytics-thop        2.0.17>=2.0.0

### Minimal Reproducible Example

```
from ultralytics.models.sam import SAM2DynamicInteractivePredictor

# Example configuration
CONF = 0.01
IMGSZ = 1024
model_path = "sam2_t.pt"

# Create predictor with max_obj_num=1
overrides = dict(conf=CONF, task="segment", mode="predict", model=model_path, save=False, imgsz=IMGSZ)
predictor = SAM2DynamicInteractivePredictor(overrides=overrides, max_obj_num=1)

# Example image + point prompt
img_file = "example.jpg"
points = [[100, 200]]
labels = [1]

# This call throws an error
results = predictor(
    source=str(img_file),
    points=[points],
    labels=[labels],
    obj_ids=[1],  # also fails with [0]
    update_memory=True
)
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @pittforce, thank you for your detailed report and for sharing an MRE 🚀! This is an automated response — an Ultralytics engineer will review and assist you here soon.

In the meantime, please help us by confirming a few details to speed up triage:
- Exact error message and full stack trace you observe 📄
- The precise commit/hash of ultralytics you’re running or `ultralytics.__version__`
- Whether the issue reproduces with different `obj_ids` values (e.g., `[0]`, `[1]`) and without `update_memory`
- A small sample image (or a synthetic placeholder) that reproduces the failure
- Any local modifications to `models/sam/predict.py` or a clean install

We also recommend verifying on the latest package and a clean environment using the steps below.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for the clear reproduction snippet and suggested location. Once you share the additional details above and confirm on the latest release, we’ll investigate further 🔍

**Comment by glenn-jocher** [comment]
Thanks for the clear repro and suggestion—confirmed. This is a 0/1-index mismatch in `SAM2DynamicInteractivePredictor` when `max_obj_num=1`: IDs are initialized 1..N, but `update_memory()` asserts 0..N-1 and `_obj_id_to_idx()` doesn’t normalize the single-object case. Minimal fix:

```python
# ultralytics/models/sam/predict.py

# In SAM2DynamicInteractivePredictor.update_memory(...)
for i, obj_id in enumerate(obj_ids):
    # normalize and validate IDs
    if self._max_obj_num == 1 and obj_id in (0, 1):
        obj_id = 1
    assert 1 <= obj_id <= self._max_obj_num, f"obj_id must be in [1, {self._max_obj_num}]"
    obj_idx = self._obj_id_to_idx(int(obj_id))
```

```python
# Also ensure _obj_id_to_idx handles the single-object and legacy 0-ID case
def _obj_id_to_idx(self, obj_id: int) -> int:
    if self._max_obj_num == 1 and obj_id in (0, 1):
        return 0
    return self.obj_id_to_idx.get(obj_id)  # raises downstream if invalid
```

Workaround until this lands: use `max_obj_num=2` and pass `obj_ids=[1]` (the extra slot stays unused). For reference on expected `obj_ids` usage, see the examples in the SAM 2 dynamic interactive section of our docs: [SAM2DynamicInteractivePredictor usage](https://docs.ultralytics.com/models/sam-2/).

If you’re open to a PR, please keep the change scoped to the above and add a small test for `max_obj_num=1`; we’ll review.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #22529: Fix object ID indexing with SAM2DynamicInteractivePredictor](https://github.com/ultralytics/ultralytics/pull/22529).

What changed and why this may resolve the issue:
- Standardizes interactive tracking to 0-based object IDs, aligning with Python conventions and removing off-by-one pitfalls.
- Refactors internal mappings to a simple identity map for `0..max_obj_num-1`, eliminating previous conversion logic.
- Updates docs and examples to use 0-based `obj_ids` for clarity and consistency.

Credit to the author and contributors:
- Authored by Y-T-G with contributions from glenn-jocher. Thank you for the improvements!

How to test the fix:
- Install the latest development version from main:
  - `pip install git+https://github.com/ultralytics/ultralytics.git@main`
- Or wait for the next official PyPI release and then upgrade:
  - `pip install -U ultralytics`
- After updating, ensure your code uses 0-based `obj_ids`. Example:
  ```python
  from ultralytics.models.sam2 import SAM2DynamicInteractivePredictor

  predictor = SAM2DynamicInteractivePredictor(max_obj_num=5)
  # Use 0-based IDs (e.g., 0, 1, 2)
  predictor.add_point(obj_id=0, x=120, y=200, label=1)
  predictor.add_point(obj_id=1, x=340, y=180, label=1)
  result = predictor.track(frame)
  ```

Feedback request:
- Please try the updated version and share whether the change resolves the issue in your environment.

Thanks for reporting this, and feel free to provide more details or logs if any problems persist. 🙏
